src/check/constrain/generate/operation.rs

```diff
@@ -334,7 +324,9 @@ fn impl_bool_op(
         &Expected::new(ast.pos, &ty),
     );
 
-    generate(left, env, ctx, constr)?;
-    generate(right, env, ctx, constr)?;
-    Ok(env.clone())
+    bin_op(left, right, env, ctx, constr)
+}
+
+fn bin_op(left: &AST, right: &AST, env: &Environment, ctx: &Context, constr: &mut ConstrBuilder) -> Constrained {
```

Comment:
Is this still necessary?
It shouldn't matter whether we are in define more or not here, right?
Variables shouldn't leak over from one side to the other.

Isn't this just a remnant from a different (wrong) solution?
The solution to our problem, it appears, was treating tuples differently by propagating out the environment when in define mode above.